### PR TITLE
Add FormObject pattern to create Conversion Projects

### PR DESCRIPTION
**Pull request by lozette**
## Changes

Note that this is not "wired in" yet and is a standalone concept. 

Inspired by https://github.com/DFE-Digital/dfe-complete-conversions-transfers-and-changes/pull/321

Add the FormObject design pattern for creating Voluntary/Involuntary Conversion Projects.

This piece of work came about when trying to lift some of the Project attributes out of the Project object and into the Conversion::Details objects. Difficulties in doing that refactor prompted us to consider adding a FormObject pattern to both create the new Conversion Projects, as well as their Conversion Details. The intention is that eventually the `create` forms will post to these FormObjects, and the FormObjects will handle the creation of all the other parts. 

Currently this is all handled in the controllers.

## Checklist

- [ ] Attach this pull request to the appropriate card in Trello.
- [ ] Update the `CHANGELOG.md` if needed.

### CODE REVIEW

**Review by beeLorna — APPROVED** [review]
I have left a few questions, but nothing needs to stop it merging if it needs to.

File: spec/support/shared_examples/create_conversion_project_form.rb [hunk 1]
```diff
@@ -11,22 +11,22 @@
       it "must be in the future" do
         form = build(
           form_factory.to_sym,
-          provisional_conversion_date: (Date.today + 1.year).at_beginning_of_month
+          provisional_conversion_date: {3 => 1, 2 => 1, 1 => 2030}
```
[the remark this answers is not in the document]
Comment (lozette) [reply spec/support/shared_examples/create_conversion_project_form.rb:14]: They should be if we are posting to the FormObject from the form in the UI; they will be passed in as a hash of three values.

File: app/forms/conversion/create_project_form.rb [hunk 2]
```diff
@@ -30,4 +32,32 @@ class Conversion::CreateProjectForm
...
   validates :establishment_sharepoint_link, presence: true, url: {hostnames: SHAREPOINT_URLS}
   validates :trust_sharepoint_link, presence: true, url: {hostnames: SHAREPOINT_URLS}
+
+  def initialize(params = {})
+    super(params)
+  end
+
+  def provisional_conversion_date=(hash)
+    @provisional_conversion_date = Date.new(year_for(hash), month_for(hash), day_for(hash))
+  rescue NoMethodError
```
[the remark this answers is not in the document]
Comment (lozette) [reply app/forms/conversion/create_project_form.rb:42]: A NoMethodError will be thrown if the hash is nil, e.g. if we try to get `value[]` on a nil in one of the `*_for` methods.

File: app/forms/conversion/create_project_form.rb [hunk 3]
```diff
@@ -30,4 +32,32 @@ class Conversion::CreateProjectForm
 
   validates :establishment_sharepoint_link, presence: true, url: {hostnames: SHAREPOINT_URLS}
   validates :trust_sharepoint_link, presence: true, url: {hostnames: SHAREPOINT_URLS}
+
+  def initialize(params = {})
```
[the remark this answers is not in the document]
Comment (lozette) [reply app/forms/conversion/create_project_form.rb:36]: No, turns out I can remove it and nothing happens!!

File: spec/support/shared_examples/create_conversion_project_form.rb [hunk 4]
```diff
@@ -113,7 +113,7 @@
     end
 
     context "when the form is valid" do
-      it "returns true" do
+      it "returns a project" do
```
[the remark this answers is not in the document]
Comment (lozette) [reply spec/support/shared_examples/create_conversion_project_form.rb:116]: Yeah I think this has ended up in the wrong commit 🤦

**Pull request merged**